### PR TITLE
added torus connector

### PR DESCRIPTION
**Pull request by YZhenY**
Hey guys!

Would love to get app.tor.us integrated into snapshot, this PR includes the connector on lock

### CODE REVIEW

File: connectors/torus.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+const get = () => import(/* webpackChunkName: "torus" */ 'torus'); // v^2.0.6
+import LockConnector from '../src/connector';
+
+export default class Connector extends LockConnector {
+  async connect() {
+    let provider;
+    try {
+      const Torus = (await get()).default;
+      const torus: any = new Torus();
+      await torus.init({showTorusButton: false});
```
Comment (bonustrack) [inline connectors/torus.ts:11]:
```suggestion
      await torus.init({ showTorusButton: false });
```

File: connectors/torus.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// @ts-ignore
+const get = () => import(/* webpackChunkName: "torus" */ 'torus'); // v^2.0.6
```
Comment (bonustrack) [inline connectors/torus.ts:2]: This package doesnt seem to be the correct one see https://www.npmjs.com/package/torus

**Review by bonustrack — COMMENTED** [review]
Hi @YZhenY ! Thank you for your contribution! Sorry for late reply i've wasn't watching this repo just noticed this PR. Can you add the logo for Torus here following the same format than others (transparency and size) https://github.com/bonustrack/lock/tree/master/connectors/assets if you do the changes we can merge your PR and test it on Snapshot.

File: connectors/torus.ts (hunk 2 above)
Comment (YZhenY) [reply connectors/torus.ts]: made the change!

**Comment by YZhenY** [comment]
made the changes! lmk if theres anything esle to do

File: connectors/torus.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+export default class Connector extends LockConnector {
+  async connect() {
+    let provider;
+    try {
+      const Torus = (await get()).default;
+      const torus: any = new Torus();
+      await torus.init({ showTorusButton: false });
+      await torus.login(); 
+      provider = torus.provider;
+      return;
```
Comment (bonustrack) [inline connectors/torus.ts:14]: There shouldn't be a `return;` here, otherwise the provider will not be returned on success. 
```suggestion
```

**Comment by bonustrack** [comment]
@YZhenY I've tried Torus connector on Snapshot but got this error
image
You'll need to find a solution for this, If that make it easier i could enable Torus current implementation on Snapshot demo so it's easier for you to debug.

**Comment by bonustrack** [comment]
You can use this PR / branch to test https://github.com/balancer-labs/snapshot/pull/121

**Comment by YZhenY** [comment]
thanks for creating the branch @bonustrack, managed to solve the issue. theres an odd reference in our provider right now. fixed it with a proxy. theres one more small issue we need to solve on ourside - theres issues with the ECSIGN on signing for a test vote.

will keep you updated here

**Comment by bonustrack** [comment]
@YZhenY Sound good! Let me know when you are ready to go

**Comment by YZhenY** [comment]
hey there @bonustrack, we've made the changes that should fix the voting bug that we encountered on snapshot.

unfortunately i was unable to replicate testing it on https://github.com/balancer-labs/snapshot/pull/121 due to an error with an external alchemy API  
image

did something change with the main build? would appreciate some help testing here

**Comment by bonustrack** [comment]
@YZhenY Hey sorry for late answer just noticed your update, we are not using Alchemy anymore so the previous issue may be solved but i've tested and got others issues, when i try to get the network or the account from the provider the requested does not resolve. Any idea what's wrong? I'm using something like this:
```
await provider.getNetwork(),
// Or 
await provider.listAccounts()
```

**Comment by YZhenY** [comment]
@bonustrack hmmm are those functions attached by code later on? they aren't part of the provider specification/ arent on metamask https://eips.ethereum.org/EIPS/eip-1193. how can i replicate this to help debug

**Comment by bonustrack** [comment]
@YZhenY Hey yes, when the provider is returned from the `connect()` function it's attached to this: https://github.com/balancer-labs/snapshot/blob/develop/src/store/modules/web3.ts#L54
Then when i call getNetwork or listAccounts from this it doesn't work. Can you try something like this:
```
import { Web3Provider } from '@ethersproject/providers';

(async function() {
  const provider = await torusConnector.connect(); // The provider returned from Torus
  const web3 = new Web3Provider(provider);
  const network = await provider.getNetwork();
  console.log(network);
})();

**Comment by YZhenY** [comment]
Hmmmm so if we call the same functions on the instance thats given back it works: 

```
 const  Web3Provider  = require('@ethersproject/providers')
 const Torus  = require("@toruslabs/torus-embed");

const torus = new Torus();

torus.init().then(
    res => {
        torus.login().then(async res => {
            const provider = torus.provider
            const web3 = new ethers.providers.Web3Provider(provider);
            console.log(web3)
            const network = await web3.getNetwork();
            console.log(network);
        }) 
    }
)```

but not on the provider directly, (which is the same in https://github.com/balancer-labs/snapshot/blob/develop/src/store/modules/web3.ts#L54) i think the above code block was a typo.

Seems like perhaps just a small bug, should be able to settle this within a 10 min call. Could we hop on a quick call at your convenience? I'm @zenzhenyu as well on telegram

**Comment by bonustrack** [comment]
@YZhenY I'm contacting you on Telegram my username is @bonustrack

**Pull request merged**